File: src/adapters/grokImageAdapter.ts
/**
 * Grok / xAI Images MCOP Adapter — wires xAI's OpenAI-compatible
 * `/v1/images/generations` endpoint into the deterministic MCOP triad.
 *
 * Companion to {@link ./grokAdapter | `GrokMCOPAdapter`} (chat-completions).
 * Kept as a separate class because the request/response shapes, retry
 * semantics, and provenance metadata differ from chat — and because the
 * Python `mcop_package/mcop/adapters/grok_image_adapter.py` already
 * follows the same split.  Sharing a base class would force one of the
 * two surfaces to grow vestigial fields.
 *
 * Like `GrokMCOPAdapter`, this adapter does NOT bundle the vendor SDK;
 * it accepts a thin `GrokImageClient` so callers can supply either the
 * bundled `defaultGrokImageClient(...)` or a fixture/replay harness.
 */

import {
  BaseAdapter,
  BaseAdapterDeps,
  PreparedDispatch,
} from './baseAdapter';
import {
  AdapterCapabilities,
  AdapterRequest,
} from './types';

/* ------------------------------------------------------------------ */
/*  Platform-native option shapes                                     */

/** xAI image-generation models known at the time of writing (May 2026). */
export type GrokImageModel =
  | 'grok-imagine-image'
  | 'grok-imagine-image-quality'
  | 'grok-imagine-image-pro'
  | (string & {});

export type GrokImageResponseFormat = 'url' | 'b64_json';

export type GrokImageAspectRatio =
  | '1:1'
  | '3:4'
  | '4:3'
  | '9:16'
  | '16:9'
  | '2:3'
  | '3:2';

export type GrokImageResolution = '1k' | '2k';

export interface GrokImageOptions {
  model?: GrokImageModel;
  /** Number of images to generate (1–10). Defaults to 1. */
  n?: number;
  responseFormat?: GrokImageResponseFormat;
  aspectRatio?: GrokImageAspectRatio;
  resolution?: GrokImageResolution;
  /** Optional caller/user identifier for vendor-side abuse monitoring. */
  user?: string;
  /** Per-request retry policy for xAI 429/5xx responses. */
  retry?: Partial<GrokImageRetryConfig>;
  /** Free-form xAI request fields not yet typed by this adapter. */
  extraBody?: Record<string, unknown>;
}

export interface GrokImageRequest extends AdapterRequest {
  payload?: {
    options?: GrokImageOptions;
  };
}

export interface GrokImageDatum {
  readonly url?: string;
  readonly b64_json?: string;
  readonly revisedPrompt?: string;
}

export interface GrokImageResult {
  readonly model: string;
  readonly responseFormat: GrokImageResponseFormat;
  readonly images: ReadonlyArray<GrokImageDatum>;
  readonly raw?: unknown;
}

export interface GrokImageRetryConfig {
  readonly maxRetries: number;
  readonly baseDelayMs: number;
  readonly maxDelayMs: number;
  readonly retryStatuses: ReadonlyArray<number>;
}

export interface GrokImagePipelineHookContext {
  readonly dispatch: PreparedDispatch;
  readonly request: GrokImageRequest;
  readonly options: GrokImageOptions;
}

export interface GrokImagePipelineHooks {
  readonly beforeDispatch?: (context: GrokImagePipelineHookContext) => void | Promise<void>;
  readonly afterDispatch?: (context: GrokImagePipelineHookContext & { result: GrokImageResult }) => void | Promise<void>;
}

/**
 * Minimal client surface — keeps the adapter SDK-agnostic and trivial to
 * mock.  Implementations MUST forward `prompt` verbatim and return one
 * `GrokImageDatum` per generated image.
 */
export interface GrokImageClient {
  generateImage(args: {
    prompt: string;
    options: GrokImageOptions;
  }): Promise<GrokImageResult>;
}

export interface GrokImageAdapterConfig extends BaseAdapterDeps {
  client: GrokImageClient;
  hooks?: GrokImagePipelineHooks;
  defaultModel?: GrokImageModel;
  /** Default entropy target for visual prompts (lower than text). */
  defaultEntropyTarget?: number;
}

export const GROK_IMAGE_PRODUCTION_PROFILE = Object.freeze({
  id: 'grok_imagine',
  adapter: 'xai-grok-images',
  defaultModel: 'grok-imagine-image' as GrokImageModel,
  qualityModel: 'grok-imagine-image-quality' as GrokImageModel,
  proModel: 'grok-imagine-image-pro' as GrokImageModel,
  entropyTarget: 0.12,
  retry: Object.freeze({
    maxRetries: 3,
    baseDelayMs: 500,
    maxDelayMs: 10_000,
    retryStatuses: [429, 500, 502, 503, 504] as ReadonlyArray<number>,
  }),
});

export class GrokImageMCOPAdapter extends BaseAdapter<
  GrokImageRequest,
  GrokImageResult
> {
  private readonly client: GrokImageClient;
  private readonly defaultModel: GrokImageModel;
  private readonly defaultEntropyTarget: number;
  private readonly hooks: GrokImagePipelineHooks;

  constructor(config: GrokImageAdapterConfig) {
    super(config);
    this.client = config.client;
    this.defaultModel = config.defaultModel ?? GROK_IMAGE_PRODUCTION_PROFILE.defaultModel;
    this.defaultEntropyTarget = config.defaultEntropyTarget ?? GROK_IMAGE_PRODUCTION_PROFILE.entropyTarget;
    this.hooks = config.hooks ?? {};
  }

  protected platformName(): string {
    return 'xai-grok-images';
  }

  async getCapabilities(): Promise<AdapterCapabilities> {
    return {
      platform: 'xai-grok-images',
      version: '2026-05',
      models: [
        'grok-imagine-image',
        'grok-imagine-image-quality',
        'grok-imagine-image-pro',
      ],
      supportsAudit: true,
      maxResolution: '2k',
      features: [
        'text-to-image',
        'batch-generation',
        'url-output',
        'base64-output',
        'aspect-ratio',
        'resolution',
        'mcop-triad-refinement',
        'human-veto',
        'merkle-audit',
      ],
      notes:
        "OpenAI-compatible Images API on https://api.x.ai/v1/images/generations. " +
        "Refined prompts route through the MCOP triad (encode → resonate → " +
        "dialectical synth → etch) before dispatch; provenance is recorded " +
        "for replay.",
    };
  }

  /**
   * Convenience facade that mirrors `GrokMCOPAdapter.generateOptimizedCompletion`.
   * Wraps the prompt in an `imaging`-domain MCOP request and returns the
   * full `AdapterResponse` carrying both the generated image data and the
   * Merkle-rooted provenance bundle.
   */
  async generateOptimizedImage(
    prompt: string,
    options: GrokImageOptions = {},
    extras: Pick<
      GrokImageRequest,
      'styleContext' | 'humanFeedback' | 'metadata' | 'entropyTarget'
    > = {},
  ) {
    return this.generate({
      prompt,
      domain: 'graphic',
      entropyTarget: extras.entropyTarget ?? this.defaultEntropyTarget,
      styleContext: extras.styleContext,
      humanFeedback: extras.humanFeedback,
      metadata: {
        ...(extras.metadata ?? {}),
        assetKind: 'image',
        model: options.model ?? this.defaultModel,
        n: options.n ?? 1,
        responseFormat: options.responseFormat ?? 'url',
      },
      payload: { options },
    });
  }

  protected async callPlatform(
    dispatch: PreparedDispatch,
    request: GrokImageRequest,
  ): Promise<GrokImageResult> {
    const opts: GrokImageOptions = {
      ...(request.payload?.options ?? {}),
      model: request.payload?.options?.model ?? this.defaultModel,
      n: clampN(request.payload?.options?.n ?? 1),
      responseFormat: request.payload?.options?.responseFormat ?? 'url',
    };

    const context: GrokImagePipelineHookContext = { dispatch, request, options: opts };
    await this.hooks.beforeDispatch?.(context);
    const result = await this.client.generateImage({
      prompt: dispatch.refinedPrompt,
      options: opts,
    });
    await this.hooks.afterDispatch?.({ ...context, result });
    return result;
  }
}

function clampN(n: number): number {
  if (!Number.isFinite(n)) return 1;
  return Math.max(1, Math.min(10, Math.floor(n)));
}

/* --------------------------------------------------------------------- */
/* Default fetch-based GrokImageClient implementation                     */
/* --------------------------------------------------------------------- */

export interface DefaultGrokImageClientConfig {
  /** xAI API key — read from `process.env.XAI_API_KEY` if absent. */
  apiKey?: string;
  /** Override the API base. Defaults to `https://api.x.ai/v1`. */
  baseUrl?: string;
  fetchImpl?: typeof fetch;
  /** Request timeout in ms. Defaults to 120s (image gen is slow). */
  timeoutMs?: number;
  retry?: Partial<GrokImageRetryConfig>;
}

interface XaiImagesResponse {
  readonly model?: string;
  readonly data: ReadonlyArray<{
    readonly url?: string;
    readonly b64_json?: string;
    readonly revised_prompt?: string;
  }>;
}

export function defaultGrokImageClient(
  config: DefaultGrokImageClientConfig = {},
): GrokImageClient {
  const apiKey =
    config.apiKey ??
    (typeof process !== 'undefined'
      ? process.env?.XAI_API_KEY
      : undefined);
  if (!apiKey || apiKey.trim().length === 0) {
    throw new Error(
      'defaultGrokImageClient: missing XAI_API_KEY — pass `apiKey` ' +
        'explicitly or set the environment variable before constructing ' +
        'the client.',
    );
  }
  const baseUrl = (config.baseUrl ?? 'https://api.x.ai/v1').replace(/\/+$/u, '');
  const fetchImpl = config.fetchImpl ?? globalThis.fetch;
  if (typeof fetchImpl !== 'function') {
    throw new Error(
      'defaultGrokImageClient: no fetch implementation available — ' +
        'provide `fetchImpl` explicitly when running on a runtime ' +
        'without a global fetch.',
    );
  }
  const timeoutMs = config.timeoutMs ?? 120_000;
  const retry = normalizeImageRetryConfig(config.retry);

  return {
    async generateImage({ prompt, options }) {
      const body: Record<string, unknown> = {
        model: options.model ?? GROK_IMAGE_PRODUCTION_PROFILE.defaultModel,
        prompt,
        n: clampN(options.n ?? 1),
        response_format: options.responseFormat ?? 'url',
      };
      if (options.aspectRatio) body.aspect_ratio = options.aspectRatio;
      if (options.resolution) body.resolution = options.resolution;
      if (options.user) body.user = options.user;
      if (options.extraBody) Object.assign(body, options.extraBody);

      const requestRetry = normalizeImageRetryConfig({
        ...retry,
        ...options.retry,
      });

      let response = await postXaiImage(fetchImpl, `${baseUrl}/images/generations`, apiKey, body, timeoutMs);
      for (let attempt = 0; shouldRetryXaiImage(response.status, attempt, requestRetry); attempt += 1) {
        const delay = computeImageBackoffMs(attempt, requestRetry);
        await sleep(delay);
        response = await postXaiImage(fetchImpl, `${baseUrl}/images/generations`, apiKey, body, timeoutMs);
      }

      if (!response.ok) {
        const detail = await safeReadText(response);
        throw new GrokImageApiError(
          `xAI image request failed: ${response.status} ${response.statusText}` +
            (detail ? ` — ${detail}` : ''),
          response.status,
          response.statusText,
          detail,
        );
      }

      const json = (await response.json()) as XaiImagesResponse;
      if (!Array.isArray(json.data)) {
        throw new Error('xAI image response did not include a data array');
      }
      const images: GrokImageDatum[] = json.data.map((d) => ({
        url: d.url,
        b64_json: d.b64_json,
        revisedPrompt: d.revised_prompt,
      }));

      return {
        model: json.model ?? options.model ?? GROK_IMAGE_PRODUCTION_PROFILE.defaultModel,
        responseFormat: options.responseFormat ?? 'url',
        images,
        raw: json,
      };
    },
  };
}

async function postXaiImage(
  fetchImpl: typeof fetch,
  url: string,
  apiKey: string,
  body: Record<string, unknown>,
  timeoutMs: number,
): Promise<Response> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    return await fetchImpl(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify(body),
      signal: controller.signal,
    });
  } finally {
    clearTimeout(timer);
  }
}

export class GrokImageApiError extends Error {
  constructor(
    message: string,
    readonly status: number,
    readonly statusText: string,
    readonly detail: string | null,
  ) {
    super(message);
    this.name = 'GrokImageApiError';
  }
}

function normalizeImageRetryConfig(
  config: Partial<GrokImageRetryConfig> = {},
): GrokImageRetryConfig {
  return {
    maxRetries: Math.max(0, Math.floor(config.maxRetries ?? GROK_IMAGE_PRODUCTION_PROFILE.retry.maxRetries)),
    baseDelayMs: Math.max(1, Math.floor(config.baseDelayMs ?? GROK_IMAGE_PRODUCTION_PROFILE.retry.baseDelayMs)),
    maxDelayMs: Math.max(1, Math.floor(config.maxDelayMs ?? GROK_IMAGE_PRODUCTION_PROFILE.retry.maxDelayMs)),
    retryStatuses: config.retryStatuses ?? GROK_IMAGE_PRODUCTION_PROFILE.retry.retryStatuses,
  };
}

function shouldRetryXaiImage(
  status: number,
  attempt: number,
  config: GrokImageRetryConfig,
): boolean {
  return attempt < config.maxRetries && config.retryStatuses.includes(status);
}

function computeImageBackoffMs(
  attempt: number,
  config: GrokImageRetryConfig,
): number {
  return Math.min(config.maxDelayMs, config.baseDelayMs * 2 ** attempt);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function safeReadText(response: Response): Promise<string | null> {
  try {
    const text = await response.text();
    return text.length > 0 ? text.slice(0, 512) : null;
  } catch {
    return null;
  }
}

File: mcop_package/mcop/adapters/grok_image_adapter.py
"""Grok Images API adapter for MCOP-Framework-2.0.

The module exposes two layers:

* ``generate_image`` / ``generate_image_async``: thin HTTP helpers for xAI's
  OpenAI-compatible Images API.
* ``GrokImageMCOPAdapter``: a BaseMCOPAdapter integration that routes image
  prompts through MCOP's encode → resonance → dialectical synthesis → etch
  pipeline before dispatching to xAI.
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Any, Dict, List, Literal, Optional, Protocol

import httpx

from .base_adapter import (
    AdapterCapabilities,
    AdapterRequest,
    AdapterResponse,
    BaseMCOPAdapter,
    PreparedDispatch,
)

GROK_BASE_URL = os.getenv("GROK_BASE_URL", "https://api.x.ai/v1")
GROK_IMAGE_MODEL = os.getenv("GROK_IMAGE_MODEL", "grok-imagine-image")

ResponseFormat = Literal["url", "b64_json"]
Resolution = Literal["1k", "2k"]
AspectRatio = Literal[
    "1:1",
    "3:4",
    "4:3",
    "9:16",
    "16:9",
    "2:3",
    "3:2",
]


__all__ = [
    "GROK_BASE_URL",
    "GROK_IMAGE_MODEL",
    "GrokImageClient",
    "GrokImageMCOPAdapter",
    "GrokImageRequest",
    "GrokImageResult",
    "generate_image",
    "generate_image_async",
]


def _api_key(explicit: Optional[str] = None) -> str:
    """Resolve the xAI bearer token from explicit args or environment."""

    key = explicit or os.getenv("GROK_API_KEY") or os.getenv("XAI_API_KEY")
    if not key:
        raise RuntimeError(
            "Grok Images API key is required; set GROK_API_KEY or XAI_API_KEY"
        )
    return key


def _endpoint(base_url: Optional[str] = None) -> str:
    return f"{(base_url or GROK_BASE_URL).rstrip('/')}/images/generations"


def _payload(
    *,
    prompt: str,
    model: str,
    n: int,
    response_format: ResponseFormat,
    aspect_ratio: Optional[AspectRatio] = None,
    resolution: Optional[Resolution] = None,
    extra_body: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    if not isinstance(prompt, str) or not prompt.strip():
        raise ValueError("prompt must be a non-empty string")
    if not 1 <= n <= 10:
        raise ValueError("n must be between 1 and 10")
    if response_format not in {"url", "b64_json"}:
        raise ValueError("response_format must be 'url' or 'b64_json'")

    payload: Dict[str, Any] = {
        "model": model,
        "prompt": prompt,
        "n": n,
        "response_format": response_format,
    }
    if aspect_ratio is not None:
        payload["aspect_ratio"] = aspect_ratio
    if resolution is not None:
        payload["resolution"] = resolution
    if extra_body:
        payload.update(extra_body)
    return payload


def _headers(api_key: Optional[str] = None) -> Dict[str, str]:
    return {
        "Authorization": f"Bearer {_api_key(api_key)}",
        "Content-Type": "application/json",
    }


def _data_from_response(response: httpx.Response) -> List[Dict[str, Any]]:
    response.raise_for_status()
    body = response.json()
    data = body.get("data")
    if not isinstance(data, list):
        raise RuntimeError("Grok Images API response did not include a data list")
    return data


def generate_image(
    prompt: str,
    *,
    model: str = GROK_IMAGE_MODEL,
    n: int = 1,
    response_format: ResponseFormat = "url",
    aspect_ratio: Optional[AspectRatio] = None,
    resolution: Optional[Resolution] = None,
    api_key: Optional[str] = None,
    base_url: Optional[str] = None,
    timeout: float = 60.0,
    client: Optional[httpx.Client] = None,
    extra_body: Optional[Dict[str, Any]] = None,
) -> List[Dict[str, Any]]:
    """Generate images with xAI's Images API and return ``data`` entries.

    Each returned dict contains either ``url`` or ``b64_json`` depending on
    ``response_format``. Pass ``client`` to reuse an existing ``httpx.Client``
    or to inject a mock transport during tests.
    """

    payload = _payload(
        prompt=prompt,
        model=model,
        n=n,
        response_format=response_format,
        aspect_ratio=aspect_ratio,
        resolution=resolution,
        extra_body=extra_body,
    )
    request_kwargs = {
        "json": payload,
        "headers": _headers(api_key),
        "timeout": timeout,
    }
    if client is not None:
        return _data_from_response(client.post(_endpoint(base_url), **request_kwargs))

    with httpx.Client() as http_client:
        return _data_from_response(
            http_client.post(_endpoint(base_url), **request_kwargs)
        )


async def generate_image_async(
    prompt: str,
    *,
    model: str = GROK_IMAGE_MODEL,
    n: int = 1,
    response_format: ResponseFormat = "url",
    aspect_ratio: Optional[AspectRatio] = None,
    resolution: Optional[Resolution] = None,
    api_key: Optional[str] = None,
    base_url: Optional[str] = None,
    timeout: float = 60.0,
    client: Optional[httpx.AsyncClient] = None,
    extra_body: Optional[Dict[str, Any]] = None,
) -> List[Dict[str, Any]]:
    """Async variant of :func:`generate_image`."""

    payload = _payload(
        prompt=prompt,
        model=model,
        n=n,
        response_format=response_format,
        aspect_ratio=aspect_ratio,
        resolution=resolution,
        extra_body=extra_body,
    )
    request_kwargs = {
        "json": payload,
        "headers": _headers(api_key),
        "timeout": timeout,
    }
    if client is not None:
        return _data_from_response(await client.post(_endpoint(base_url), **request_kwargs))

    async with httpx.AsyncClient() as http_client:
        return _data_from_response(
            await http_client.post(_endpoint(base_url), **request_kwargs)
        )


class GrokImageClient(Protocol):
    """Minimal sync client surface for Grok image generation."""

    def generate_image(
        self,
        *,
        prompt: str,
        model: str,
        n: int,
        response_format: ResponseFormat,
        aspect_ratio: Optional[AspectRatio] = None,
        resolution: Optional[Resolution] = None,
        extra_body: Optional[Dict[str, Any]] = None,
    ) -> List[Dict[str, Any]]:  # pragma: no cover -- structural typing only
        ...


@dataclass
class GrokImageResult:
    """Vendor-neutral result returned by ``GrokImageMCOPAdapter``."""

    model: str
    images: List[Dict[str, Any]]
    response_format: ResponseFormat
    audit_hash: Optional[str] = None
    raw: Any = None

    @property
    def image_url(self) -> Optional[str]:
        first = self.images[0] if self.images else {}
        value = first.get("url")
        return value if isinstance(value, str) else None


@dataclass
class GrokImageRequest(AdapterRequest):
    """MCOP request shape for xAI image generations."""

    model: str = GROK_IMAGE_MODEL
    n: int = 1
    response_format: ResponseFormat = "url"
    aspect_ratio: Optional[AspectRatio] = None
    resolution: Optional[Resolution] = None


class _DefaultGrokImageClient:
    def generate_image(
        self,
        *,
        prompt: str,
        model: str,
        n: int,
        response_format: ResponseFormat,
        aspect_ratio: Optional[AspectRatio] = None,
        resolution: Optional[Resolution] = None,
        extra_body: Optional[Dict[str, Any]] = None,
    ) -> List[Dict[str, Any]]:
        return generate_image(
            prompt,
            model=model,
            n=n,
            response_format=response_format,
            aspect_ratio=aspect_ratio,
            resolution=resolution,
            extra_body=extra_body,
        )


class GrokImageMCOPAdapter(BaseMCOPAdapter):
    """MCOP adapter that dispatches refined prompts to xAI Images."""

    def __init__(
        self,
        client: Optional[GrokImageClient] = None,
        **base_kwargs: Any,
    ) -> None:
        super().__init__(**base_kwargs)
        self._client = client or _DefaultGrokImageClient()

    @property
    def platform_name(self) -> str:
        return "grok-images"

    def get_capabilities(self) -> AdapterCapabilities:
        return AdapterCapabilities(
            platform="grok-images",
            version="2026-05",
            models=[
                "grok-imagine-image",
                "grok-imagine-image-quality",
                "grok-imagine-image-fast",
            ],
            supports_audit=True,
            features=[
                "text-to-image",
                "batch-generation",
                "url-output",
                "base64-output",
                "aspect-ratio",
                "resolution",
                "merkle-audit",
            ],
            max_resolution="2k",
            notes=(
                "Uses xAI's /images/generations endpoint; set "
                "GROK_API_KEY or XAI_API_KEY for the default HTTP client."
            ),
        )

    def generate_image(
        self,
        prompt: str,
        *,
        model: str = GROK_IMAGE_MODEL,
        n: int = 1,
        response_format: ResponseFormat = "url",
        aspect_ratio: Optional[AspectRatio] = None,
        resolution: Optional[Resolution] = None,
        **metadata: Any,
    ) -> AdapterResponse:
        request = GrokImageRequest(
            prompt=prompt,
            domain="imaging",
            model=model,
            n=n,
            response_format=response_format,
            aspect_ratio=aspect_ratio,
            resolution=resolution,
            metadata=metadata,
        )
        return self.generate(request)

    def call_platform(
        self, dispatch: PreparedDispatch, request: AdapterRequest
    ) -> GrokImageResult:
        if isinstance(request, GrokImageRequest):
            image_request = request
        else:
            payload = request.payload
            image_request = GrokImageRequest(
                prompt=request.prompt,
                domain=request.domain,
                metadata=request.metadata,
                human_feedback=request.human_feedback,
                style_context=request.style_context,
                entropy_target=request.entropy_target,
                planned_sequence=request.planned_sequence,
                model=str(payload.get("model", GROK_IMAGE_MODEL)),
                n=int(payload.get("n", 1)),
                response_format=payload.get("response_format", "url"),
                aspect_ratio=payload.get("aspect_ratio"),
                resolution=payload.get("resolution"),
            )

        images = self._client.generate_image(
            prompt=dispatch.refined_prompt,
            model=image_request.model,
            n=image_request.n,
            response_format=image_request.response_format,
            aspect_ratio=image_request.aspect_ratio,
            resolution=image_request.resolution,
            extra_body=None,
        )
        return GrokImageResult(
            model=image_request.model,
            images=images,
            response_format=image_request.response_format,
            audit_hash=dispatch.etch_hash or None,
            raw={"data": images},
        )
